### PR TITLE
Add experimental "pedantic" fortio output formatter.

### LINKED ISSUE #514: Nighthawk fortio's output is discrepant then real fortio's output
**Description**
Nighthawk's fortio transform output is different than real fortio output for a few fields.

#### Nighthawk's Output
```json
{
 "Labels": "",
 "RequestedQPS": 5,
 "ActualQPS": 4.9999995930000329,
 "ActualDuration": 5000000407,
 "NumThreads": 100,
 "DurationHistogram": {
  "Count": "25",
  "Data": [
   {
    "Start": 0.024088575,
    "End": 0.024088575,
    "Percent": 0,
    "Count": "1"
   },
```

#### Fortio's Output
```json
{
  "RunType": "HTTP",
  "Labels": "test",
  "StartTime": "2020-09-08T16:29:19.35507+05:30",
  "RequestedQPS": "100",
  "RequestedDuration": "5s",
  "ActualQPS": 59.9799880886016,
  "ActualDuration": 5101701580,
  "NumThreads": 8,
  "Version": "1.6.8",
  "DurationHistogram": {
    "Count": 306,
    "Min": 0.114543168,
    "Max": 0.436206912,
    "Sum": 40.422087638,
    "Avg": 0.1320983256143791,
    "StdDev": 0.03333137761133907,
    "Data": [
      {
        "Start": 0.114543168,
        "End": 0.2,
        "Percent": 97.38562091503267,
        "Count": 298
      },
```

As can be seen 
- RequestedQPS is an integer but should be a string.
- Dataentry.count, Sizes.count, HeaderSizes.count is a string but should be an integer.

**Expected Behaviour**

Nighthawk Output should be similar to Fortio's Output.

### PR DESCRIPTION
**Pull request by oschaaf**
Adds a post-processing step to massage the fortio output formatting
to more exactly reproduce Fortio's output, while leaving the
original output formatter intact for backwards compatibility purposes.

Fixes known issues #422 and #514. Experimental until we have
some confirmation that this can be marked as final / all issues have
been resolved.

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by dubious90** [comment]
@eric846 Please review and assign back to me once done.

**Review by kushthedude — DISMISSED** [review]
This works for us!
/cc @leecalcote

**Comment by oschaaf** [comment]
Also /cc @carolynhu 
this is pulled of in a backwards compatible manner, but when we upgrade, there'll
be a way to get rid of the workarounds for the fortio output more quirks in NH once this lands.

File: source/client/output_formatter_impl.cc
Comment (carolynhu) [inline source/client/output_formatter_impl.cc:403]: typo here? `RequestedQPS`

File: test/test_data/output_formatter.medium.fortio-noquirks.gold
Comment (carolynhu) [inline test/test_data/output_formatter.medium.fortio-noquirks.gold:201]: Could we move these fields (line 197 to line 201) up to under line 8 "Count" field as what fortio has?

File: test/test_data/output_formatter.medium.fortio-noquirks.gold
Comment (carolynhu) [inline test/test_data/output_formatter.medium.fortio-noquirks.gold:475]: are these value fake or real? as this is in seconds...

**Comment by leecalcote** [comment]
@oschaaf thank you for today's call. Very much enjoyed it.

We can assist in verifying compatibility. Are the build artifacts of this PR (the container image) available?

File: test/test_data/output_formatter.medium.fortio-noquirks.gold
Comment (oschaaf) [reply test/test_data/output_formatter.medium.fortio-noquirks.gold:201]: That is quite hard to pull off in the current approach. Looking at it, the proto json serialiser we use here alphabetically sorts this. I think it can do that, because after parsing the difference usually doesn't matter? Do you see a way that the re-ordering could cause trouble?

File: test/test_data/output_formatter.medium.fortio-noquirks.gold
Comment (oschaaf) [reply test/test_data/output_formatter.medium.fortio-noquirks.gold:475]: I didn't come up with this test data, it's a copy of a pre-existing one modified to reflect the changes in here. If I interpret the file correctly, we're not looking at seconds, but rather a histogram of byte counts for the header?

File: source/client/output_formatter_impl.cc
Comment (oschaaf) [reply source/client/output_formatter_impl.cc]: Fix in 5ab48900

**Comment by oschaaf** [comment]
> @oschaaf thank you for today's call. Very much enjoyed it.
> 
> We can assist in verifying compatibility. Are the build artifacts of this PR (the container image) available?

@leecalcote 
Unfortunately, I think fresh images will only be pushed post-merge of this. 
So to try this out one needs to manually build the docker image. 
Having said that, it is ok to iterate on this while we have the feature marked as --experimental, as we don't
need to be backwards compatible while that is the case. So waiting until after the merge is viable too.

File: test/test_data/output_formatter.medium.fortio-noquirks.gold
Comment (carolynhu) [reply test/test_data/output_formatter.medium.fortio-noquirks.gold:201]: Not yet. Just noticed the order is mismatch with fortio. Don't think it will cause any trouble.

**Review by eric846 — DISMISSED** [review]
LGTM

File: source/client/output_formatter_impl.cc
Comment (mum4k) [inline source/client/output_formatter_impl.cc:404]: (optional) Would you mind including (in the comment) and example of the matched string? That usually helps when trying to understand regexes in code. This one is fairly trivial, so optional, but we could use one for the second regex.

File: source/client/output_formatter_impl.h
Comment (mum4k) [inline source/client/output_formatter_impl.h:111]: Can we add a comment on the class roughly explaining what it is for?

File: test/output_formatter_test.cc
Comment (mum4k) [inline test/output_formatter_test.cc:229]: The "auto" here doesn't seem to hide redundant type information, so it might be better for the reader to spell it out.

File: test/output_formatter_test.cc
Comment (mum4k) [inline test/output_formatter_test.cc:231]: This golden file is fairly long which makes it hard for the reader to figure out which portions actually are relevant to this test and which are just noise. Any way we could cut it down to the minimum required to test this new behavior?

**Review by mum4k — CHANGES_REQUESTED** [review]

File: source/client/output_formatter_impl.cc
Comment (oschaaf) [reply source/client/output_formatter_impl.cc:404]: Yeah, I always felt that regular expressions are kind of a write-only concept :-)
Done in 2df8c2be

File: source/client/output_formatter_impl.h
Comment (oschaaf) [reply source/client/output_formatter_impl.h:119]: Done in 2df8c2be

File: test/output_formatter_test.cc
Comment (oschaaf) [reply test/output_formatter_test.cc]: Done here and in other spots in 2df8c2be

File: test/output_formatter_test.cc
Comment (oschaaf) [reply test/output_formatter_test.cc:234]: This makes sense, but is kind of pre-existing; I think the initial work for this just grabbed a raw arbitrary snapshot and included it for testing the origin fortio output formatter code. 
We are now more or less forced to propagate that, as our new gold file is an amended version of the original one. 
I punted this to #542 to track doing this, do that work for you?

Comment (mum4k) [reply test/output_formatter_test.cc:234]: That works, thank you.

**Review by mum4k — APPROVED** [review]

**Pull request merged**
